lib/IRGen/GenFunc.cpp

```diff
@@ -2433,3 +2434,83 @@ IRGenFunction::createAsyncDispatchFn(const FunctionPointer &fnPtr,
   Builder.CreateRetVoid();
   return dispatch;
 }
+
+void IRGenFunction::emitSuspensionPoint(llvm::Value *toExecutor,
+                                        llvm::Value *asyncResume) {
+  // TODO: pointerauth
+
+  // Setup the suspend point.
+  SmallVector<llvm::Value *, 8> arguments;
+  arguments.push_back(asyncResume);
+  auto resumeProjFn = getOrCreateResumeFromSuspensionFn();
+  arguments.push_back(
+      Builder.CreateBitOrPointerCast(resumeProjFn, IGM.Int8PtrTy));
+  llvm::Function *suspendFn = createAsyncSuspendFn();
+  arguments.push_back(
+      Builder.CreateBitOrPointerCast(suspendFn, IGM.Int8PtrTy));
+
+  arguments.push_back(asyncResume);
+  arguments.push_back(
+      Builder.CreateBitOrPointerCast(toExecutor, getAsyncExecutor()->getType()));
+  arguments.push_back(getAsyncTask());
+  arguments.push_back(getAsyncExecutor());
+  arguments.push_back(getAsyncContext());
+
+  emitSuspendAsyncCall(arguments);
```

Comment:
Could you emit a fast-path branch around this that avoids the suspension point when the executors are exactly equal?

Reply:
Yes. Erik and I have discussed this as a follow-up optimization.